### PR TITLE
added material.from_library method

### PR DESCRIPTION
**Pull request by shimwell**
Hi all,

This PR attempts to add the ability to add material composition and density from a predefined collection / library.

This PR follows on from [this](https://openmc.discourse.group/t/add-pnnl-15870-to-openmc-model/1134/4) discussion where adding materials from the [PNNL](https://www.pnnl.gov/main/publications/external/technical_reports/PNNL-15870Rev2.pdf) library of predefined materials was discussed.

Adding a PNNL_v1 material database has previously be performed by [PyNe](https://github.com/pyne/pyne/) and more recently by [ATTILA](https://silverfirsoftware.com/announcements/). I've also done something similar before in the [neutronics-material-maker](https://github.com/fusion-energy/neutronics_material_maker) package.

One of the differences to existing implementations and this PR is that we have revision version 2 (released in 2021) of the PNNL compendium released in here instead of the first revision (released in 2011)

I think this method for this implementation is quite concise due to expanding the dictionary into keyword arguments with ``**`` . The implementation is also extendable as the JSON file can support multiple libraries if PNNL_v3 gets release or there is a need to add another material compendium.

I've also added a simple test to the test_material.py unit test but the usage is simple 

```python
>>>import openmc

>>>m1 = openmc.Material()
>>>m1.add_from_library(name='Sodium Oxide', library='pnnl_v2')
>>>m1
Material
	ID             =	1
	Name           =	
	Temperature    =	None
	Density        =	2.27 [g/cc]
	S(a,b) Tables  
	Nuclides       
	O16            =	0.332523     [ao]
	O17            =	0.0001266665 [ao]
	O18            =	0.0006833327 [ao]
	Na23           =	0.666667     [ao]
```

### CODE REVIEW

**Comment by eepeterson** [comment]
@shimwell I really like this! My first bit of feedback is that I think having a single json file for each library may be clearer and I think the method should be a class method on the `Material` class so something more like this: `m1 = openmc.Material.from_library(name='Sodium Oxide', library='pnnl_v2')`

**Comment by shimwell** [comment]
Thanks for the feedback @eepeterson I've moved the method to a classmethod as suggested.

The python example from the first comment now changes to

```python
>>>import openmc

>>>m1 = openmc.Material.from_library(name='Sodium Oxide', library='pnnl_v2')
>>>m1
Material
	ID             =	1
	Name           =	
	Temperature    =	None
	Density        =	2.27 [g/cc]
	S(a,b) Tables  
	Nuclides       
	O16            =	0.332523     [ao]
	O17            =	0.0001266665 [ao]
	O18            =	0.0006833327 [ao]
	Na23           =	0.666667     [ao]
```

**Comment by shimwell** [comment]
@eepeterson I've refactored the PR to use a single files for each library as suggested :+1: . I guess this will be handy when adding private material libraries.

**Review by paulromano — COMMENTED** [review]
Thanks for the contribution @shimwell! One question I have is how the JSON file is being generated. I see that in the material compendium in many cases they have elemental compositions, but in the JSON file we have nuclides explicitly. This will create problems if the data library being used doesn't have a particular nuclide. For example, Anthracene uses C12 and C13, which won't work with ENDF/B-VII.1 and earlier. Is it possible to switch over to using elemental compositions when available?

**Comment by paulromano** [comment]
@shimwell Also looks like you have a conflict on material.py

**Comment by eepeterson** [comment]
@paulromano and @shimwell I have been thinking about this a little lately and I think this might warrant just a little more structure to do robustly, namely a `MaterialLibrary` class with to/from file methods, the ability to handle elemental compositions in addition to nuclide compositions, as well as the ability to merge existing `materials.xml` files into a library, and nice printing/repr for QA purposes. I also think an optional environment variable like `OPENMC_MATERIAL_LIBRARY_PATH` as a colon separated list of paths (like `$PATH`) would be powerful and separate any internally stored libraries like the PNNL_v2 one in this PR from other custom libraries. All of these features are pretty straightforward, the only question in my mind is if we want the JSON format for libraries or if there is another preferred format. Thoughts?

**Comment by shimwell** [comment]
I was just looking at a nice PR https://github.com/openmc-dev/openmc/pull/2139 . Once #2139 is merged in then this PR could be refactored to make use of the new methods introduced. The JSON file could also be reduced in side considerably to dictionaries of tuples instead of the current rather verbose dictionary.

File: openmc/data/data.py [hunk 1]
```diff
@@ -204,6 +204,9 @@
 _HALF_LIFE = {}
 _LOG_TWO = log(2.0)
 
+_INTERNAL_DATA_PATH = Path(__file__).parent
```
Comment (eepeterson) [inline openmc/data/data.py:207]:
```suggestion
INTERNAL_DATA_PATH = Path(__file__).parent
```

**Review by eepeterson — CHANGES_REQUESTED** [review]
first round of small comments

File: openmc/data/data.py [hunk 2]
```diff
@@ -204,6 +204,9 @@
...
+
```
Comment (shimwell) [inline openmc/data/data.py:208]:
```suggestion
INTERNAL_DATA_PATH = Path(__file__).parent

```

File: openmc/material.py [hunk 3]
```diff
@@ -1210,6 +1212,59 @@ def mix_materials(cls, materials, fracs: typing.Iterable[float],
...
+                    f'Supported libraries are {available_mat_libraries}'
+            )
+            raise ValueError(msg)
+
+        global MATERIAL_LIBRARIES
+
+        # loads in the library into the MATERIAL_LIBRARIES if not already loaded
+        if library not in MATERIAL_LIBRARIES.keys():
+            mat_lib_path = openmc.data.INTERNAL_DATA_PATH / f'material_library_{library}.json'
+            MATERIAL_LIBRARIES[library] = json.loads(mat_lib_path.read_text())
```
Comment (eepeterson) [inline openmc/material.py:1245]:
```suggestion
        try:
            library_data = _MATERIAL_LIBRARY_CACHE[library]
        except KeyError:
            library_data = json.loads(MATERIAL_LIBRARIES[library].read_text())
            _MATERIAL_LIBRARY_CACHE[library] = library_data
```

File: openmc/material.py [hunk 4]
```diff
@@ -1210,6 +1212,59 @@ def mix_materials(cls, materials, fracs: typing.Iterable[float],
...
+
+        if name not in MATERIAL_LIBRARIES[library].keys():
+            msg = (
+                    f'material name {name} not found in the {library} materials '
+                    f'library. Available are {list(MATERIAL_LIBRARIES[library].keys())}'
```
Comment (eepeterson) [inline openmc/material.py:1250]:
```suggestion
        if name not in library_data.keys():
            msg = (
                    f'material name {name} not found in the {library} materials '
                    f'library. Available are {list(library_data.keys())}'
```

File: openmc/material.py [hunk 5]
```diff
@@ -1210,6 +1212,59 @@ def mix_materials(cls, materials, fracs: typing.Iterable[float],
...
+            )
+            raise ValueError(msg)
+
+        material_to_add = MATERIAL_LIBRARIES[library][name]
```
Comment (eepeterson) [inline openmc/material.py:1254]:
```suggestion
        material_to_add = library_data[name]
```

**Review by eepeterson — CHANGES_REQUESTED** [review]
I did a little refactoring here so that we can use external libraries as well based on an environment variable

**Comment by eepeterson** [comment]
@shimwell I think we need to move the library file to `material_libraries/pnnl_v2.json` to fix the failing test

**Comment by shimwell** [comment]
> @shimwell I think we need to move the library file to `material_libraries/pnnl_v2.json` to fix the failing test

Thanks Ethan. I've created the folder, renamed the file and moved it in there. I've also added it to the package data in setup.py. Hoping the tests pass this time :eyes:

File: openmc/material.py [hunk 6]
```diff
@@ -24,6 +25,25 @@
...
                  'macro')
 
+_MAT_LIB_DIRS = [openmc.data.INTERNAL_DATA_PATH / 'material_libraries']
+if os.environ.get('OPENMC_MATERIAL_LIBRARY_PATH') is not None:
+    _MAT_LIB_DIRS += [Path(a) for a in os.environ.get('OPENMC_MATERIAL_LIBRARY_PATH').split(':')]
+
+_MAT_LIB_PATHS = [f.resolve() for p in _MAT_LIB_DIRS
+                  for f in sorted(Path(p).glob('*'))]
+
+MATERIAL_LIBRARIES = {}
```
Comment (pshriwise) [inline openmc/material.py:35]: The difference between the types of values in `MATERIAL_LIBRARIES` and `_MATERIAL_LIBRARY_CACHE` wasn't clear to me until reading the code. Can we add some comments on the types of keys and values in each?

File: openmc/material.py [hunk 7]
```diff
@@ -1210,6 +1230,59 @@ def mix_materials(cls, materials, fracs: typing.Iterable[float],
...
+            )
+            raise ValueError(msg)
+            
+        global _MATERIAL_LIBRARY_CACHE
+
+        # loads in the library into the MATERIAL_LIBRARIES if not already loaded
+        try:
+            library_data = _MATERIAL_LIBRARY_CACHE[library]
+        except KeyError:
+            library_data = json.loads(MATERIAL_LIBRARIES[library].read_text())
```
Comment (pshriwise) [inline openmc/material.py:1262]: It looks like we are loading in the material library from the JSON file "as-is". This is fine w/ me but it means we expect those JSON files to have a particular structure -- did we document this somewhere for users who want to create a their own, custom library?

File: tests/unit_tests/test_material.py [hunk 8]
```diff
@@ -496,3 +496,19 @@ def test_activity_of_metastable():
     m1.set_density('g/cm3', 1)
     m1.volume = 98.9
     assert pytest.approx(m1.activity, rel=0.001) == 1.93e19
+
+
+def test_material_from_library():
+    """Checks that the material is added from the pnnl_v2 library it is assigned
```
Comment (pshriwise) [inline tests/unit_tests/test_material.py:502]:
```suggestion
    """Checks that the material is added from the pnnl_v2 library and that it is assigned
```

**Review by pshriwise — CHANGES_REQUESTED** [review]
This is a really handy capability. I spent a good chunk of time translating material definitions from the PNNL compendium by hand when learning to use MCNP. Thanks @shimwell and @eepeterson for working on this!

Just a few small thoughts/comments from me.

File: openmc/material.py (hunk 7 above)
Comment (eepeterson) [reply openmc/material.py:1262]: I would probably personally prefer to have a well defined material library format in XML for consistency with other data (especially materials). We could almost just use the material XML format but ignore ID numbers. Although there is the complication of not expanding elements so it is a bit different, but for validating a spec and roundtrip consistency I might go with XML - thoughts?

File: openmc/material.py [hunk 9]
```diff
@@ -1210,6 +1230,59 @@ def mix_materials(cls, materials, fracs: typing.Iterable[float],
...
+
+        mat = cls()
+        if 'nuclides' in material_to_add.keys():
+            for nuclide in material_to_add['nuclides']:
+                mat.add_nuclide(**nuclide)
+        if 'elements' in material_to_add.keys():
+            for nuclide in material_to_add['elements']:
+                mat.add_element(**nuclide)
+
+        mat.set_density(**material_to_add['density'])
```
[the remark this answers is not in the document]
Comment (eepeterson) [reply openmc/material.py:1282]: probably, yes

File: openmc/material.py (hunk 6 above)
Comment (eepeterson) [inline openmc/material.py:35]:
```suggestion
# MATERIAL_LIBRARIES maps strings of library names to pathlib.Path objects for 
# each available material library
MATERIAL_LIBRARIES = {}
```

File: openmc/material.py [hunk 10]
```diff
@@ -24,6 +25,25 @@
...
 
+_MAT_LIB_DIRS = [openmc.data.INTERNAL_DATA_PATH / 'material_libraries']
+if os.environ.get('OPENMC_MATERIAL_LIBRARY_PATH') is not None:
+    _MAT_LIB_DIRS += [Path(a) for a in os.environ.get('OPENMC_MATERIAL_LIBRARY_PATH').split(':')]
+
+_MAT_LIB_PATHS = [f.resolve() for p in _MAT_LIB_DIRS
+                  for f in sorted(Path(p).glob('*'))]
+
+MATERIAL_LIBRARIES = {}
+_MATERIAL_LIBRARY_CACHE = {}
```
Comment (eepeterson) [inline openmc/material.py:36]:
```suggestion
# _MATERIAL_LIBRARY_CACHE maps strings of library names to the loaded library to
# prevent loading the same library multiple times or loading unused libraries
_MATERIAL_LIBRARY_CACHE = {}
```

File: openmc/material.py (hunk 6 above)
Comment (eepeterson) [reply openmc/material.py:37]: done

File: openmc/material.py (hunk 7 above)
Comment (shimwell) [reply openmc/material.py:1266]: XML files would mean we can use the ```Material.from_xml_element``` method if the materials are defined entirely as nuclides (no elements)

**Comment by shimwell** [comment]
Having a bit of a chat on slack and we are keen to change the underlying file format to XML.

We are also keen to keep the data file small (which requires elemental descriptions of the materials)

To proceed with this PR we need to add support for elements to the materials.XML file format and convert the JSON file on this PR to an XML file.

This PR will go on hold for a little time while we add element support to the materials.xml file

File: openmc/material.py (hunk 7 above)
Comment (shimwell) [reply openmc/material.py]: PR is now using xml files as suggested

File: openmc/material.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (shimwell) [reply openmc/material.py]: materials are now taken from and xml element and if this includes temperature then it will be in the material

File: openmc/material.py [hunk 11]
```diff
@@ -3,6 +3,7 @@
 from copy import deepcopy
 from numbers import Real
 from pathlib import Path
+import json
```
Comment (shimwell) [inline openmc/material.py:6]:
```suggestion
```

File: setup.py [hunk 12]
```diff
@@ -33,7 +33,8 @@
     'package_data': {
         'openmc.lib': ['libopenmc.{}'.format(suffix)],
         'openmc.data': ['mass16.txt', 'BREMX.DAT', 'half_life.json', '*.h5'],
-        'openmc.data.effective_dose': ['*.txt']
+        'openmc.data.effective_dose': ['*.txt'],
+        'openmc.data.material_libraries': ['*.json']
```
Comment (shimwell) [inline setup.py:37]:
```suggestion
        'openmc.data.material_libraries': ['*.xml']
```

**Comment by shimwell** [comment]
I came across a little error that the CI didn't catch.

when using ```pip install .``` the pnnl_v2.xml file was not being copied over to the directory where python package was installed.

So I've added an ```__init__.py``` file to the openmc/data/material_libraries folder that helps the setup.py realize that these xml files should be copied

File: openmc/material.py [hunk 13]
```diff
@@ -1275,6 +1298,55 @@ def mix_materials(cls, materials, fracs: typing.Iterable[float],
...
+                    f'material name {name} not found in the {library} materials '
+                    f'library. Available are {list(library_data.keys())}'
+            )
+            raise ValueError(msg)
+
+        material_to_add = library_data[name]
+
+        # mat = cls()
+        
+
```
Comment (shimwell) [inline openmc/material.py:1347]:
```suggestion
```

**Comment by shimwell** [comment]
I have left this PR for a while as the plan was to refactor the materials class so that is accepts elements then come back to this PR.

However I have an alternative suggestions.

How about we remove the ```openmc/data/material_libraries/pnnl_v2.xml``` file and the code that adds the internal library so we are no longer adding a specific PNNL library to the code. However we retain the functionality of adding a users xml material library file?

basically just retaining part of this PR so that users can load in their own material library.

This idea was inspired by @mwart-cfs who has a use case I had not considered. For QA purposes they prefer no material libraries to be initially loaded into the code but want the ability to add user libraries. Further details [here](https://github.com/fusion-energy/neutronics_material_maker/pull/28)

**Comment by eepeterson** [comment]
I don’t quite see how not including widely used libraries within OpenMC directly helps with QA exactly and to me it doesn’t make sense for something like the PNNL compendium to exist within every users custom libraries. Since properly including elements in material specifications does require significant refactoring, maybe it is worth implementing a temporary way of including user libraries, but im not sure.

**Comment by shimwell** [comment]
How about we break this PR into two parts.

Part 1, just the frame work to allow users to load in their own custom libraries (xml files)

Part 2, after the material refactor add in the PNNL compendium so that it is distributed with openmc package and available as a default loaded in library

**Comment by shimwell** [comment]
I think this might be worth another go as the materials compendium is now downloadable in json format https://compendium.cwmd.pnnl.gov/

**Comment by MicahGale** [comment]
> I came across a little error that the CI didn't catch. Perhaps because we use `pip install -e .` which links but doesn't copy files
> 
> when using `pip install .` the pnnl_v2.xml file was not being copied over to the directory where python package was installed.
> 
> So I've added an `__init__.py` file to the openmc/data/material_libraries folder that helps the setup.py realize that these xml files should be copied

I think this is a deprecated approach. The more `pyproject` way I think would be through `MANIFEST.in`

**Comment by MicahGale** [comment]
> I don’t quite see how not including widely used libraries within OpenMC directly helps with QA exactly and to me it doesn’t make sense for something like the PNNL compendium to exist within every users custom libraries. Since properly including elements in material specifications does require significant refactoring, maybe it is worth implementing a temporary way of including user libraries, but im not sure.

If an organizational user doesn't want to use the libraries due to QA, they can just exclude that from the scope of their QA program.

**Comment by shimwell** [comment]
This PR has been replaced
